src/Type/Definition/TypeWithFields.php

```diff
@@ -0,0 +1,94 @@
+<?php
+
+declare(strict_types=1);
+
+namespace GraphQL\Type\Definition;
+
+use GraphQL\Error\InvariantViolation;
+use GraphQL\Utils\Utils;
+
+use function array_keys;
+
+abstract class TypeWithFields extends Type implements HasFieldsType
+{
+    /**
+     * Lazily initialized.
+     *
+     * @var array<string, FieldDefinition>
+     */
+    private array $fields;
+
+    private function initializeFields(): void
+    {
+        if (isset($this->fields)) {
+            return;
+        }
+
+        $fields       = $this->config['fields'] ?? [];
+        $this->fields = FieldDefinition::defineFieldMap($this, $fields);
+    }
+
+    /**
+     * @throws InvariantViolation
+     */
+    public function getField(string $name): FieldDefinition
+    {
+        Utils::invariant($this->hasField($name), 'Field "%s" is not defined for type "%s"', $name, $this->name);
+
+        return $this->findField($name);
+    }
+
+    public function findField(string $name): ?FieldDefinition
+    {
+        $this->initializeFields();
+
+        if (! isset($this->fields[$name])) {
+            return null;
+        }
+
+        if ($this->fields[$name] instanceof UnresolvedFieldDefinition) {
+            $this->fields[$name] = $this->fields[$name]->resolve();
+        }
+
+        return $this->fields[$name];
+    }
+
+    public function hasField(string $name): bool
+    {
+        $this->initializeFields();
+
+        return isset($this->fields[$name]);
+    }
+
+    /**
+     * @return array<string, FieldDefinition>
+     *
+     * @throws InvariantViolation
+     */
```

Comment:
I'd use this 👇🏾  where comments are useful but are in interface already

```suggestion
    /** @inheritDoc */
```

Reply:
Ah yeah, interestingly with this docblock:

```
    /**
     * @throws InvariantViolation
     */
```

`phpcs` removes the `@inhertiDoc` (it leaves it where there is also an `@return`)... that is the `@inheritDoc` even needed?

Reply:
No. It's needed only for array types IIRC. Otherwise it's considered to be already inherited automatically from the interface without the need to specify the inheritance.